### PR TITLE
Remove legacy dependencies from PSA tests

### LINKED ISSUE #7061: Remove legacy dependencies from PSA tests
PSA test suites should mostly not have legacy dependencies anymore. We mass-removed them with the help of [`tests/scripts/set_psa_test_dependencies.py`](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/set_psa_test_dependencies.py) in #4012, #4158 and #4246. But we don't run this script on a regular basis. Some legacy dependencies have either escaped the renaming (some negative test cases aren't handled by the script) or snuck back in.

The goal of this task is to remove legacy dependencies from `test_suite_psa_*` except for ones that we decide are explicitly excused. The following dependencies are excluded from this issue:

* `MBEDTLS_GENPRIME` when generating RSA keys — no PSA equivalent.
* `MBEDTLS_ENTROPY_NV_SEED` in init tests — no PSA equivalent.
* `MBEDTLS_PK_WRITE_C` and `MBEDTLS_PK_PARSE_C` — tracked in https://github.com/Mbed-TLS/mbedtls/issues/7055.

Definition of done: this should return no hits (currently 135):
```
perl -ne 's/depends_on:// or next; print "$ARGV:$.:$_\n" foreach grep !/^MBEDTLS_(GENPRIME|PSA_.*|PK_PARSE_C|PK_WRITE_C|ENTROPY_NV_SEED)$/, /\bMBEDTLS_\w+/g' tests/suites/test_suite_psa*
```

Note: if some hits really can't be translated to PSA, remove the ones that can be, and evaluate whether the remaining hits should be excused or what's going on.

Although not a dependency, we'll probably want to do https://github.com/Mbed-TLS/mbedtls/issues/7062 around the same time, to avoid having to deal with the problem again in the future.

### LINKED ISSUE #7062: Forbid legacy dependencies in PSA tests
PSA test suites should mostly not have legacy dependencies anymore. We mass-removed them with the help of [`tests/scripts/set_psa_test_dependencies.py`](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/set_psa_test_dependencies.py) in #4012, #4158 and #4246. But we don't run this script on a regular basis. Some legacy dependencies have either escaped the renaming (some negative test cases aren't handled by the script) or snuck back in.

The goal of this task is to enforce the absence of `MBEDTLS_xxx` dependencies in `tests/suites/test_suite_psa_*.data` and `tests/suites/test_suite_psa_*.function`. Extend `component_check_test_cases` in `all.sh`.

Exceptions which have no PSA equivalent:

* `MBEDTLS_GENPRIME` when generating RSA keys.
* `MBEDTLS_ENTROPY_NV_SEED` in init tests.
* `MBEDTLS_PSA_xxx` for parts of PSA (e.g. driver support, storage support).

Prerequisites:

* [ ] https://github.com/Mbed-TLS/mbedtls/issues/7061
* [ ] https://github.com/Mbed-TLS/mbedtls/issues/7055

### PR DESCRIPTION
**Pull request by mpg**
## Description

Resolves #7061 and resolves #7062.

## PR checklist

- [x] **changelog** not required - test changes only
- [x] **backport** not required - we only care in development
- [x] **tests** provided

### CODE REVIEW

**Review by valeriosetti — APPROVED** [review]
I like the test component added to `all.sh`!
The PR looks OK to me. OpenCI also fully passed, so I think this is fine

File: tests/scripts/all.sh [hunk 1]
```diff
@@ -1024,6 +1024,43 @@ component_check_test_cases () {
...
 }
 
+component_check_test_dependencies () {
+    msg "Check: test case dependencies: legacy vs PSA" # < 1s
+
+    found="check-test-deps-found-$$"
+    expected="check-test-deps-expected-$$"
+
+    # Find legacy dependencies in PSA tests
+    grep 'depends_on' tests/suites/test_suite_psa* |
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:1034]: The wildcard pattern might include things we don't want in developer scripts, such as editor backup files.

```suggestion
    grep 'depends_on' tests/suites/test_suite_psa*.data tests/suites/test_suite_psa*.function |
```

File: tests/scripts/all.sh [hunk 2]
```diff
@@ -1024,6 +1024,43 @@ component_check_test_cases () {
...
+    echo "MBEDTLS_ECP_RESTARTABLE" >> $expected
+    # No PSA equivalent - needed by some init tests
+    echo "MBEDTLS_ENTROPY_NV_SEED" >> $expected
+    # Used by two tests that are about an extension to the PSA standard;
+    # as such, no PSA equivalent.
+    echo "MBEDTLS_PEM_PARSE_C" >> $expected
+
+    # Compare reality with expectation.
+    # We want an exact match, to ensure the above list remains up-to-date.
+    diff -U0 $expected $found
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:1059]: If this fails on the CI, the output will be something like

```
******************************************************************
* check_test_dependencies: Check: test case dependencies: legacy vs PSA 
* Thu Nov  9 16:09:10 CET 2023
******************************************************************
--- check-test-deps-expected-1228447    2023-11-09 16:09:10.409940159 +0100
+++ check-test-deps-found-1228447       2023-11-09 16:09:10.409940159 +0100
@@ -4,0 +5 @@
+MBEDTLS_OOPS
```

Is this comprehensible enough for someone who's not deeply familiar with the problem we want to solve here?

It's annoying that this code doesn't remember where the forbidden identifiers were found. But admittedly that would be more work and I'm not sure it's justified.

File: tests/scripts/all.sh [hunk 3]
```diff
@@ -1024,6 +1024,43 @@ component_check_test_cases () {
...
+    found="check-test-deps-found-$$"
+    expected="check-test-deps-expected-$$"
+
+    # Find legacy dependencies in PSA tests
+    grep 'depends_on' tests/suites/test_suite_psa* |
+        grep -Eo '!?MBEDTLS_[^: ]*' |
+        grep -v MBEDTLS_PSA_ |
+        sort -u > $found
+
+    # Expected ones with justification - keep in sorted order!
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:1039]: Thanks for the detailed explanations for each exception!

Please also add some text explaining the problem we want to solve. For someone who isn't familiar with the history here, I don't think that's easy to understand.

File: tests/scripts/all.sh (hunk 2 above)
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:1059]: Please add `diff` to the list of allow-keep-going (like `grep`) in `can_keep_going_after_failure`.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM apart from some fragility in `all.sh`.

Note: I authored one of the commits here. I presume that Manuel's inclusion counts as having approved it.

File: tests/scripts/all.sh (hunk 2 above)
Comment (mpg) [reply tests/scripts/all.sh:1059]: > Is this comprehensible enough for someone who's not deeply familiar with the problem we want to solve here?
> 
Not sure. I'll see what I can do to make it better.

> It's annoying that this code doesn't remember where the forbidden identifiers were found. But admittedly that would be more work and I'm not sure it's justified.

I'm not sure it's justified. Most of the time the test should pass, and when it fails finding an unexpected occurrence of `MBEDTLS_FOO` then it's easy enough for anyone investigating the problem to `grep -n MBEDTLS_FOO tests/suites/test_suite_psa*` - that's what I did for this PR anyway.

File: tests/suites/test_suite_psa_crypto.data [hunk 4]
```diff
@@ -225,19 +225,19 @@ depends_on:PSA_WANT_ALG_RSA_PKCS1V15_SIGN:PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_BASIC:P
...
 
 PSA import RSA public key: 1022-bit (not supported)
-depends_on:MBEDTLS_RSA_C
+depends_on:PSA_WANT_KEY_TYPE_RSA_PUBLIC_KEY
```
[the remark this answers is not in the document]
Comment (mpg) [reply tests/suites/test_suite_psa_crypto.data:228]: Thanks for explaining the background!

(Btw, we're not planning to do driver-only RSA any time soon, but when I have some free time, I intend to try it out in a reduced config like `crypto_full` minus PK, where I expect it should work out of the box.)

File: tests/scripts/all.sh (hunk 2 above)
Comment (mpg) [reply tests/scripts/all.sh:1077]: > Not sure. I'll see what I can do to make it better.

I went with comments. I think if anyone sees output that doesn't make sense to them, looking at the component that generated this output is a reasonable next step.

**Review by valeriosetti — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Pull request merged**